### DISCUSSION TITLE
Is there an API for custom UI themes/elements?

### DISCUSSION THREAD

**Discussion by cafuneandchill** [Monika-After-Story/MonikaModDev#8231] [QA - code and content creation (technical)]
I like the inclusion of Night Mode, but I'm not too psyched about the colors, despite them being true to the original.

Therefore, I'd *love* to try and make something of myself, but I'm curious if there's a better way to do it than by straight up replacing the UI assets.

**Comment by multimokia** [comment]
Having been dabbling in this myself, this is something I have wanted to build as overriding can get messy especially if/when _multiple_ UI submods would get involved. So I've been wanting to think out an approach to UI customizability but also allowing submods to _not_ conflict with each other when applied.

In principle it would also allow mixing and matching UI elements for granular control or just flat out using one complete theme.

**Comment by cafuneandchill** [comment]
I think a mixed approach would be best. In this case, a UI submod would contain two primary things:

- A collection of separate UI elements with a certain theme
  - main menu background
  - buttons
  - fonts
  - etc.
- A preset that combines all these elements

Thus, a user could go two ways about customizing the UI:

1. If they wish so, they could use elements from different submods.

    For example:

    | Preset | *Custom* |
    |--------|---------|

    | UI element | Submod |
    |----------------------|-----------------------------|
    | Main Menu Background | *XMAS_UI_Submod_main_menu_BG* |
    | Buttons              | *Rounded Default Buttons*     |
    | Text Font            | *Iosevka Aile*                |

2. If the user wishes to use only a certain UI submod, then they would just pick the preset that uses all the elements included in the submod.

    For example:

    | Preset | XMAS UI Submod |
    |--------|----------|

    | UI element | Submod |
    |------------|-------------|
    | Main Menu Background | XMAS_UI_Submod_main_menu_BG |
    | Buttons | XMAS UI Submod Buttons |
    | Text Font | Christmas Cookies Regular |

I believe a number of Linux desktop environments (modern XFCE, for example) use a similar technique, where they would have Gtk themes and icons under a single theme (Papirus, Matcha, etc.), while giving an option to pick them separately.

**Comment by Booplicate** [comment]
> In principle it would also allow mixing and matching UI elements for granular control

> If they wish so, they could use elements from different submods.

Unsure if this would be worth. Yes, it'd allow to mix themes, but it's quite possible most themes wouldn't work together anyway. It's also already a non-trivial task. A theme isn't just a background image and a font. We use renpy's styles. Thus a theme consist of [multiple properties](https://www.renpy.org/doc/html/style_properties.html), and then there's also a [GUI system](https://www.renpy.org/doc/html/gui.html). So I think if somebody likes 2 themes and wants to use both, they should create a new theme and handle style conflicts themselves. I don't really see our end-users selecting and adjusting styles in a menu with 9999 options.

I'm thinking of something that would allow to map all styles as well as images and pos constants used in custom displayables. Ideally that should be done automatically, w/o having to maintain the map. That would be a map with the default styles. Then a submod could create their own map (of course via an API), which would be merged with the default one overriding the default values. After that we'd call a function to rebuild styles using the new map.

There's still a lot of problems to solve
- should dark mode be considered a theme, or should we assume every theme has light + dark modes?
- how do we get references to class properties? Could have a func for that, but then
- what to do about objects already created (during init and/or runtime)? Keep weak references?

**Comment by cafuneandchill** [comment]
> I'm thinking of something that would allow to map all styles as well as images and pos constants used in custom displayables. Ideally that should be done automatically, w/o having to maintain the map. That would be a map with the default styles. Then a submod could create their own map (of course via an API), which would be merged with the default one overriding the default values. After that we'd call a function to rebuild styles using the new map.

That sounds nice. If I were implementing that, I would also probably use a JSON or YAML file for such a map.

> should dark mode be considered a theme, or should we assume every theme has light + dark modes?

I'm probably biased, but I'm all for the former option -- make Dark Mode as a separate theme.

> - how do we get references to class properties? Could have a func for that, but then
> - what to do about objects already created (during init and/or runtime)? Keep weak references?

I have no idea how MAS is structured internally, so I don't have much say in this, but for the objects created at runtime I would try to make a separate func that would just update their state, according to the selected theme (although re-initializing them + garbage collecting old weak references sounds fine too). Either that, or just making it so that UI changes could be performed only when MAS is not up and running.

**Comment by multimokia** [comment]
>     * should dark mode be considered a theme, or should we assume every theme has light + dark modes?

Some sort of manifest or something would be best to declare whether or not a theme would contribute that.
This is why I was thinking of allowing mix/match as ideally, a theme would have to specify _what_ it intends to override since it may not be everything.

That and doing it via re-initializing styles/overriding functions/displayables can be extremely problematic in terms of MAS updating. Therefore it makes sense to allow a theme to specify what it contributes --including for its own day/night versions so MAS can understand that it has just those.

The map idea would make sense for sure

**Comment by multimokia** [comment]
That said if other themes can provide their own dark variant within the same pack. I think the official MAS dark mode can be considered as the same theme as the day, just the night version of it.

**Comment by Booplicate** [comment]
We have a nice (although, kinda laggy) system to switch styles one the fly, hence I want to keep the dark mode and light mode as one theme. But it also makes things harder (including potential styles API) and it's also really *slow*.

> a theme would have to specify what it intends to override since it may not be everything.

This is why I first thought of a map. If you don't add a style there, it'd use the default style. This is a must since it'd help with future updates. For example, we added a new displayable, but a submod hasn't been updated. That way the submod would still work reusing the default assets/styles.

> making it so that UI changes could be performed only when MAS is not up and running.

A bit inconvenient for users, but that's the easiest solution, and most apps require a restart to apply style changes. I didn't mention that only because we have a system to update styles w/o having to restart.

> I would also probably use a JSON or YAML file for such a map.

To be consistent, I'd say use `JSON` as that's the format we use in MAS.

**Comment by multimokia** [comment]
I'm not sure I like no runtime change as a lot of users would prefer to preview or test the themes and then swap to another potentially. This just becomes cumbersome if the user cannot switch without restarting.

Unsure about which applications require restarts. Discord, Word, Chrome/Firefox/Safari, Github Desktop, etc. All can change their styles on the fly.

Probably best for UX if we scale our existing system to work with more than just two styles

**Comment by Booplicate** [comment]
> Unsure about which applications require restarts. Discord, Word, Chrome/Firefox/Safari, Github Desktop, etc. All can change their styles on the fly.

I should've said most non-web-apps. if only we could use CSS.

> a lot of users would prefer to preview or test the themes and then swap to another potentially

I don't really see this because to install a theme they would need to restart. I'd assume people would look at the screenshots of the theme before downloading and installing it. But it would be nice if we could switch w/o restarting.

**Comment by ThePotatoGuy** [comment]
> I have no idea how MAS is structured internally, so I don't have much say in this, but for the objects created at runtime I would try to make a separate func that would just update their state, according to the selected theme

Dark mode is implemented in a way that already does this. The mod expects a ton of styles to have a dark suffix, and any things that don't follow that practice have custom handling in `mas_darkMode`. This is basically an "on the fly" theme change as well. Some other things use an explicit dark mode check, but that should be changed to the current system.

So the natural extension is to just make the suffix generalizable and require some sort of mapping for the custom things (as mentioned by boop).
